saga/saga.go

```diff
@@ -0,0 +1,118 @@
+package saga
+
+type SagaRecoveryType int
+
+/*
+ * Saga Recovery Types define how to interpret SagaState in RecoveryMode.
+ *
+ * ForwardRecovery: all tasks in the saga must be executed at least once.
+ *                  tasks MUST BE idempotent
+ *
+ * RollbackRecovery: if Saga is Aborted or in unsafe state, compensating
+ *                   tasks for all started tasks need to be executed.
+ *                   compensating tasks MUST BE idempotent.
+ */
+const (
+	BackwardRecovery SagaRecoveryType = iota
+	ForwardRecovery
+)
+
+/*
+ * Saga Object which provides all Saga Functionality
+ * Implementations of SagaLog should provide a factory method
+ * which returns a saga based on its implementation.
+ */
+type Saga struct {
```

Comment:
Hrm, I'm not sure this will work for implementations outside the package, because they won't be able to set log...